### PR TITLE
Pull Request - Epic 7.3 - Worksheet Generator

### PR DESCRIPTION
**Pull request by AaronSosaRamos**
# Pull Request - Epic 7.3 - Worksheet Generator
By: Synthwave Sentinels

## Summary
This PR allows educators to generate worksheets based in the grade level, topic, question types and number of question per type. As a result, we based our feature in the existent course type division made by the (IB Diploma Programme)[https://www.researchgate.net/publication/329548502_Impacts_of_International_Baccalaureate_programmes_on_teaching_and_learning_A_review_of_the_literature]. Moreover, educators can specify the context with the extended file support implemented with the following file types for a better domain-specific adaptation: 

- PDF
- CSV
- TXT
- MD
- URL (Such as Wikipedia or other source with info.)
- PPTX
- DOCX
- XLS / XLSX
- XML
- GOOGLE DOCS
- GOOGLE SHEETS
- GOOGLE SLIDES
- GOOGLE PDF
- IMAGES (PNG / JPG / JPEG)

## Changes
- Reused the document loaders from Epic 7.2 #61.
- Reused the util file `allowed_file_extensions_dynamo.py` for a robust approach in the file type management from Epic 7.2 #61.
- Created a new `tool.py` for integrating the designed workflow for generating the worksheet:
image
- Created a new core.py` for integrating the worksheet generator workflow from `tool.py` appropriately.
- Decoupling WorksheetGenerator's args. as a best practice for allowing flexible maintainability like our Epic 7.7 #73.
- Created new Error Classes for improving the error mapping process.
- Enabled a Computer Vision approach for generating woksheets from images as an innovative proposal for the project.

## Testing
Each document loader was tested using pytest (Managing appropriate scenarios and edge cases).
image

## Results
All document loaders are working appropriately with the optimal results.
image
image

## Notes

For Google Docs, Slides, Sheets and PDF files: 
- Those files need to be shared as **public** in Google Drive.
- They need to be uploaded to Google Drive. If a Google File is created in Google Drive, it needs to be downloaded and then uploaded to be detected. This was the most appropriate approach that we discovered as the GoogleDriver loader from LangChain works with OAuth2, which is not appropriate for production deployment.

## How to Test
1. Clone the repo in your local

2. Create and activate virtual environment

3. Use pip install -r requirements.txt to install required libraries.

4. Create .env file with ENV_TYPE, GCP_PROJECT_ID and GOOGLE_API_KEY fields. Env type is dev, gcp_project_id is your project id from cloud console project, google_api_key is your api key from AI studio.

5. Then type ./local-start.sh to start the application.

6. Add sample requests and responses for each file type as mentioned above in screenshots and test it.

### Schema:
```python
class WorksheetQuestionModel(BaseModel):
    question_type: str
    number: int

class WorksheetGeneratorArgs(BaseModel):
    grade_level: str
    topic: str
    worksheet_list: List[WorksheetQuestionModel]
    file_url: str
    file_type: str
    lang: Optional[str] = "en"
```

### Example:
```json
{
   "user":{
      "id":"string",
      "fullName":"string",
      "email":"string"
   },
   "type":"chat",
   "tool_data":{
      "tool_id":2,
      "inputs":[
         {
            "name":"worksheet_generator_args",
            "value":{
               "grade_level":"College",
               "topic":"Machine Learning",
               "worksheet_list":[
                  {
                     "question_type":"fill_in_the_blank",
                     "number":2
                  },
                  {
                     "question_type":"multiple_choice_question",
                     "number":2
                  }
               ],
               "file_url":"https://www.interactions.com/wp-content/uploads/2017/06/machine_learning_wp-5.pdf",
               "file_type":"pdf",
               "lang":"en"
            }
         }
      ]
   }
}
```

### Schema:
```python
#Fill-in-the-blank question type
class QuestionBlank(BaseModel):
    key: str = Field(description="A unique identifier for the blank, starting from 0.")
    value: str = Field(description="The text content to fill in the blank")

class FillInTheBlankQuestion(BaseModel):
    question: str = Field(description="The question text with blanks indicated by placeholders (It must be 5 blank spaces {0}, {1}, {2}, {3}, {4})")
    blanks: List[QuestionBlank] = Field(description="A list of blanks for the question, each with a key and a value")
    word_bank: List[str] = Field(description="A list of the correct texts that fill in the blanks, in random order")
    explanation: str = Field(description="An explanation of why the answers are correct")

    model_config = {
        "json_schema_extra": {
            "examples": """ 
                {
                "question": "The {0} of France is {1}, and it is known for its {2} and {3} {4}.",
                "blanks": [
                    {"key": "0", "value": "capital"},
                    {"key": "1", "value": "Paris"},
                    {"key": "2", "value": "art"},
                    {"key": "3", "value": "culinary"},
                    {"key": "4", "value": "delights"}
                ],
                "word_bank": ["delights", "art", "Paris", "culinary", "capital"],
                "explanation": "Paris is the capital of France, and it is renowned for its contributions to art and its exceptional culinary scene."
              }
          """
        }
    }

#Open-ended question type
class OpenEndedQuestion(BaseModel):
    question: str = Field(description="The open-ended question text")
    answer: str = Field(description="The expected correct answer")
    feedback: List[str] = Field(description="A list of possible answers for the provided question")

    model_config = {
        "json_schema_extra": {
            "examples": """ 
                {
                "question": "What is the significance of Paris in French history?",
                "answer": "Paris is the capital of France and has been a major center for politics, culture, art, and history.",
                "feedback": [
                    "Paris is the capital of France.",
                    "Paris has been a cultural center in Europe.",
                    "Paris played a major role in the French Revolution."
                ]
              }
          """
        }
    }

#True-False question type
class TrueFalseQuestion(BaseModel):
    question: str = Field(description="The True-False question text")
    answer: bool = Field(description="The correct answer, either True or False")
    explanation: str = Field(description="An explanation of why the answer is correct")

    model_config = {
        "json_schema_extra": {
            "examples": """ 
                {
                "question": "The Eiffel Tower is located in Paris.",
                "answer": true,
                "explanation": "The Eiffel Tower is a famous landmark located in Paris, France."
              }
          """
        }
    }

#Multiple Choice question type
class QuestionChoice(BaseModel):
    key: str = Field(description="A unique identifier for the choice using letters A, B, C, or D.")
    value: str = Field(description="The text content of the choice")
class MultipleChoiceQuestion(BaseModel):
    question: str = Field(description="The question text")
    choices: List[QuestionChoice] = Field(description="A list of choices for the question, each with a key and a value")
    answer: str = Field(description="The key of the correct answer from the choices list")
    explanation: str = Field(description="An explanation of why the answer is correct")

    model_config = {
        "json_schema_extra": {
            "examples": """ 
                {
                "question": "What is the capital of France?",
                "choices": [
                    {"key": "A", "value": "Berlin"},
                    {"key": "B", "value": "Madrid"},
                    {"key": "C", "value": "Paris"},
                    {"key": "D", "value": "Rome"}
                ],
                "answer": "C",
                "explanation": "Paris is the capital of France."
              }
          """
        }

      }

#Relate concepts question type
class TermMeaningPair(BaseModel):
    term: str = Field(description="The term to be matched")
    meaning: str = Field(description="The meaning of the term")

class RelateConceptsQuestion(BaseModel):
    question: str = Field(description="The 'Relate concepts' question text. It must be appropriate for generating pairs and answers.")
    pairs: List[TermMeaningPair] = Field(description="A list of term-meaning pairs in disorder")
    answer: List[TermMeaningPair] = Field(description="A list of the correct term-meaning pairs in order")
    explanation: str = Field(description="An explanation of the correct term-meaning pairs")

    model_config = {
        "json_schema_extra": {
            "examples": """ 
                {
                "question": "Match each term with its correct meaning.",
                "pairs": [
                    {
                        "term": "Chlorophyll",
                        "meaning": "The powerhouse of the cell, where respiration and energy production occur."
                    },
                    {
                        "term": "Photosynthesis",
                        "meaning": "A green pigment responsible for the absorption of light to provide energy for photosynthesis."
                    },
                    {
                        "term": "Mitochondria",
                        "meaning": "The process by which green plants use sunlight to synthesize foods with the help of chlorophyll."
                    },
                    {
                        "term": "Nucleus",
                        "meaning": "The gel-like substance inside the cell membrane."
                    },
                    {
                        "term": "Cytoplasm",
                        "meaning": "The control center of the cell that contains DNA."
                    }
                ],
                "answer": [
                    {
                        "term": "Photosynthesis",
                        "meaning": "The process by which green plants use sunlight to synthesize foods with the help of chlorophyll."
                    },
                    {
                        "term": "Chlorophyll",
                        "meaning": "A green pigment responsible for the absorption of light to provide energy for photosynthesis."
                    },
                    {
                        "term": "Mitochondria",
                        "meaning": "The powerhouse of the cell, where respiration and energy production occur."
                    },
                    {
                        "term": "Nucleus",
                        "meaning": "The control center of the cell that contains DNA."
                    },
                    {
                        "term": "Cytoplasm",
                        "meaning": "The gel-like substance inside the cell membrane."
                    }
                ],
                "explanation": "Photosynthesis involves using sunlight to create food in plants, facilitated by chlorophyll. Mitochondria are involved in energy production in cells. The nucleus is the control center of the cell, and the cytoplasm is the gel-like substance within the cell membrane."
              }
          """
        }
    }

#Math. Exercise question type
class MathExerciseQuestion(BaseModel):
    question: str = Field(description="The math exercise question text")
    solution: str = Field(description="The step-by-step solution to the math problem")
    correct_answer: str = Field(description="The correct answer to the math problem")
    explanation: str = Field(description="An explanation of why the solution is correct")

    model_config = {
        "json_schema_extra": {
            "examples": """ 
                {
                "question": "Solve the equation: 2x + 3 = 11",
                "solution": "Step 1: Subtract 3 from both sides to get 2x = 8. Step 2: Divide both sides by 2 to get x = 4.",
                "correct_answer": "4",
                "explanation": "By isolating the variable x, we find that x equals 4."
              }
          """
        }
    }
```

### Example: 
```json 
{
  "data": {
    "fill_in_the_blank": [
      {
        "question": "In {0} learning, a machine uses {1} data to discover new patterns without prior information, a process often used in {2}, where data is categorized into groups of {3} data. This contrasts with {4} learning, which relies on labeled data.",
        "blanks": [
          {
            "key": "0",
            "value": "unsupervised"
          },
          {
            "key": "1",
            "value": "unlabeled"
          },
          {
            "key": "2",
            "value": "clustering"
          },
          {
            "key": "3",
            "value": "similar"
          },
          {
            "key": "4",
            "value": "supervised"
          }
        ],
        "word_bank": [
          "supervised",
          "clustering",
          "unsupervised",
          "unlabeled",
          "similar"
        ],
        "explanation": "Unsupervised learning involves machines learning from unlabeled data to identify patterns, often used in clustering, which groups similar data. This differs from supervised learning, where machines learn from labeled datasets."
      },
      {
        "question": "In the realm of {0} learning, algorithms leverage {1} data to discern novel patterns without relying on pre-existing information. This approach proves particularly valuable in {2}, where datasets are organized into clusters based on data {3}. Conversely, {4} learning necessitates labeled data for training.",
        "blanks": [
          {
            "key": "0",
            "value": "unsupervised"
          },
          {
            "key": "1",
            "value": "unlabeled"
          },
          {
            "key": "2",
            "value": "clustering"
          },
          {
            "key": "3",
            "value": "similarity"
          },
          {
            "key": "4",
            "value": "supervised"
          }
        ],
        "word_bank": [
          "supervised",
          "unsupervised",
          "clustering",
          "unlabeled",
          "similarity"
        ],
        "explanation": "Unsupervised learning algorithms excel at identifying patterns in unlabeled data, making them well-suited for clustering tasks where data points with similarities are grouped together. This contrasts with supervised learning, which depends on labeled data for training."
      }
    ],
    "multiple_choice_question": [
      {
        "question": "In the context of supervised machine learning, how does an algorithm 'learn' to classify data accurately?",
        "choices": [
          {
            "key": "A",
            "value": "It analyzes unlabeled data to identify emerging patterns and categorize them without prior knowledge."
          },
          {
            "key": "B",
            "value": "It interacts with the environment and learns through trial and error to achieve a desired outcome."
          },
          {
            "key": "C",
            "value": "It compares its predicted outputs with a set of correct outputs, adjusting its model to minimize errors."
          },
          {
            "key": "D",
            "value": "It relies on pre-programmed rules and logic to categorize data based on predefined criteria."
          }
        ],
        "answer": "C",
        "explanation": "Supervised learning involves training an algorithm on labeled data, where it receives both inputs and corresponding correct outputs. The algorithm learns by comparing its own predictions to the correct answers, adjusting its internal model to reduce discrepancies and improve accuracy."
      },
      {
        "question": "Compare and contrast supervised and unsupervised learning techniques in machine learning, providing examples of how each is used.",
        "choices": [
          {
            "key": "A",
            "value": "Supervised learning uses labeled data to predict outcomes (e.g., classifying images), while unsupervised learning finds patterns in unlabeled data (e.g., grouping customers by purchase history)."
          },
          {
            "key": "B",
            "value": "Supervised learning relies on trial and error for optimization (e.g., game playing), while unsupervised learning uses pre-defined rules (e.g., expert systems)."
          },
          {
            "key": "C",
            "value": "Supervised learning is used for simple tasks (e.g., linear regression), while unsupervised learning handles complex problems (e.g., natural language processing)."
          },
          {
            "key": "D",
            "value": "Supervised learning is primarily used in robotics, while unsupervised learning is dominant in image recognition."
          }
        ],
        "answer": "A",
        "explanation": "Supervised learning uses labeled data to train algorithms to make predictions, such as classifying images. Unsupervised learning identifies patterns in unlabeled data, like grouping customers by purchase history, without specific output guidance."
      }
    ]
  }
}
```

### References
[Request Template](https://docs.google.com/document/d/1Jgfv2zYekrVZ9DJDkuh1uY2Pgx_jaTUVUHf_C07wEEM)
[Responses](https://docs.google.com/document/d/11GztDejIJj9cz-3iSww6Wxh0v-j5Nw7udFtgkuuz_x0)

### CODE REVIEW

File: README.md
Comment (hash2004) [inline README.md:1]: There is no need to change the README.md file. You can see the updated from the STAGING branch.

File: app/features/worksheet_generator/document_loaders.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
+from langchain_community.document_loaders import YoutubeLoader
```
Comment (hash2004) [inline app/features/worksheet_generator/document_loaders.py:1]: You can import the document loaders in a single line.

File: app/features/worksheet_generator/document_loaders.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+from langchain_community.document_loaders import PyPDFLoader
```
Comment (hash2004) [inline app/features/worksheet_generator/document_loaders.py:2]: You can import the document loaders in a single line.

File: app/features/worksheet_generator/document_loaders.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+from langchain_community.document_loaders.csv_loader import CSVLoader
+from langchain_community.document_loaders import TextLoader
```
Comment (hash2004) [inline app/features/worksheet_generator/document_loaders.py:4]: You can import the document loaders in a single line.

File: app/features/worksheet_generator/document_loaders.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+from langchain_community.document_loaders import UnstructuredURLLoader
```
Comment (hash2004) [inline app/features/worksheet_generator/document_loaders.py:5]: You can import the document loaders in a single line.

File: app/features/worksheet_generator/document_loaders.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+from langchain_community.document_loaders import UnstructuredPowerPointLoader
```
Comment (hash2004) [inline app/features/worksheet_generator/document_loaders.py:6]: You can import the document loaders in a single line.

File: app/features/worksheet_generator/metadata.json
Comment (hash2004) [inline app/features/worksheet_generator/metadata.json:1]: Please write more readable descriptions against label, name and type.

File: app/features/worksheet_generator/prompts/generate-topic-prompt.txt
Comment (hash2004) [inline app/features/worksheet_generator/prompts/generate-topic-prompt.txt:1]: Use /n line breaks to specify where lines are breaking. LLMs perform best when you use prompting techniques like CoT, few shot and more. Here is a resource: https://www.promptingguide.ai/techniques

File: app/features/worksheet_generator/prompts/generate-worksheet-prompt.txt
Comment (hash2004) [inline app/features/worksheet_generator/prompts/generate-worksheet-prompt.txt:1]: Use /n line breaks to specify where lines are breaking. LLMs perform best when you use prompting techniques like CoT, few shot and more. Here is a resource: https://www.promptingguide.ai/techniques

**Review by hash2004 — CHANGES_REQUESTED** [review]
Great work! Please make the changes and get back to us :)

File: README.md
Comment (AaronSosaRamos) [reply README.md:1]: Done!

File: app/features/worksheet_generator/document_loaders.py (hunk 1 above)
Comment (AaronSosaRamos) [reply app/features/worksheet_generator/document_loaders.py]: Done!

File: app/features/worksheet_generator/document_loaders.py (hunk 2 above)
Comment (AaronSosaRamos) [reply app/features/worksheet_generator/document_loaders.py]: Done!

File: app/features/worksheet_generator/document_loaders.py (hunk 3 above)
Comment (AaronSosaRamos) [reply app/features/worksheet_generator/document_loaders.py]: Done!

File: app/features/worksheet_generator/document_loaders.py (hunk 4 above)
Comment (AaronSosaRamos) [reply app/features/worksheet_generator/document_loaders.py]: Done!

File: app/features/worksheet_generator/document_loaders.py (hunk 5 above)
Comment (AaronSosaRamos) [reply app/features/worksheet_generator/document_loaders.py]: Done!

**Review by yunusj — APPROVED** [review]

**Pull request merged**